clap_man/src/lib.rs

```diff
@@ -0,0 +1,132 @@
+#![doc(html_logo_url = "https://raw.githubusercontent.com/clap-rs/clap/master/assets/clap.png")]
+#![doc = include_str!("../README.md")]
+#![warn(missing_docs, trivial_casts, unused_allocation, trivial_numeric_casts)]
+#![forbid(unsafe_code)]
+#![deny(missing_docs)]
+
+mod render;
+
+pub use roff;
+
+use render::subcommand_heading;
+use roff::{roman, Roff};
+use std::io::Write;
+
+/// Generate a manual page and write it out.
+pub fn generate_manpage<'a>(app: &mut clap::App<'a>, buf: &mut dyn Write) -> Result<(), std::io::Error> {
+    let meta = Meta::from_clap("1", "", app);
+    let man = Man::new(&meta, app);
+    man.render(buf)
+}
+
+/// Metadata about a manual page.
+pub struct Meta {
+    title: String,
+    section: String,
+    date: String,
+    source: String,
+    manual: String,
+}
+
+impl Meta {
+    /// Create metadata from a clap::App.
+    pub fn from_clap(section: &str, manual: &str, app: &clap::App) -> Self {
+        Self {
+            title: app.get_name().to_string(),
+            section: section.to_string(),
+            date: "".to_string(), // FIXME
+            source: format!(
+                "{} {}",
+                app.get_name(),
+                app.get_version().unwrap_or_default()
+            ),
+            manual: manual.to_string(),
+        }
+    }
+
+    // Turn metadata into arguments for a .TH macro.
+    fn to_args(&self) -> Vec<&str> {
+        vec![
+            &self.title,
+            &self.section,
+            &self.date,
+            &self.source,
+            &self.manual,
+        ]
+    }
+}
+
+/// A manual page as constructed from a clap::App.
+pub struct Man {
+    roff: Roff,
+}
+
+impl Man {
+    /// Create a new manual page.
+    pub fn new(meta: &Meta, app: &mut clap::App) -> Self {
```

Comment:
Why do we have `Meta`?  Seems like a weird extra layer

Reply:
Would you prefer a builder-pattern for the meta information? The section, source and manual can probably be mostly automatic but the date and title should probably be user configurable.